src/jingle_sip/jingle_sip_callbacks.erl

```diff
@@ -0,0 +1,276 @@
+%% @doc
+%% This module defines callbacks for nksip application
+%% Exported functions whill be called when there is new message to MongooseIM
+%% or when there is an response to a SIP INVITE sent from MongooseIM to a SIP Proxy
+-module(jingle_sip_callbacks).
+
+-include("mongoose.hrl").
+-include("jlib.hrl").
+-include_lib("nksip/include/nksip.hrl").
+-include_lib("nksip/include/nksip_call.hrl").
+
+%% this is because nksip has wrong type specs
+-dialyzer({nowarn_function, [sip_invite_unsafe/2,
+                             sip_reinvite_unsafe/2,
+                             sip_bye/2,
+                             sip_cancel/3,
+                             send_ringing_session_info/1,
+                             invite_resp_callback/1
+                             ]}).
+
+%% SIP callbacks
+-export([sip_invite/2]).
+-export([sip_reinvite/2]).
+-export([sip_info/2]).
+-export([sip_bye/2]).
+-export([sip_cancel/3]).
+-export([sip_dialog_update/3]).
+-export([invite_resp_callback/1]).
+
+sip_invite(Req, Call) ->
+    try
+        sip_invite_unsafe(Req, Call)
+    catch Class:Reason ->
+            StackTrace = erlang:get_stacktrace(),
+            ?WARNING_MSG("Error parsing sip invite, class=~p, reason=~p, stacktrace=~p",
+                         [Class, Reason, StackTrace]),
+            {error, request_not_parsable}
+    end.
+
+sip_reinvite(Req, Call) ->
+    ?WARNING_MSG("re-INVITE: ~p", [Req]),
+    try
+        sip_reinvite_unsafe(Req, Call)
+    catch Class:Reason ->
+            StackTrace = erlang:get_stacktrace(),
+            ?WARNING_MSG("Error parsing sip invite, class=~p, reason=~p, stacktrace=~p",
+                         [Class, Reason, StackTrace]),
+            {error, request_not_parsable}
+    end.
+
+
+sip_invite_unsafe(Req, _Call) ->
+    {FromJID, FromBinary} = get_user_from_sip_msg(from, Req),
+    {ToJID, ToBinary} = get_user_from_sip_msg(to, Req),
+
+    case ejabberd_sm:is_offline(ToJID) of
+        false ->
+            translate_and_deliver_invite(Req, FromJID, FromBinary, ToJID, ToBinary);
+        _ ->
+            {reply, temporarily_unavailable}
+    end.
+
+translate_and_deliver_invite(Req, FromJID, FromBinary, ToJID, ToBinary) ->
+    CallID = nksip_sipmsg:header(<<"call-id">>, Req),
+    Body = nksip_sipmsg:meta(body, Req),
+
```

Comment:
I am not a fun of empty lines inside a function.

Reply:
They sometimes do organise code to be me a bit more readable :)